examples/udev.rs

```diff
@@ -0,0 +1,472 @@
+
+    /*
+     * Initialize the udev backend
+     */
+    let context = libudev::Context::new().unwrap();
+    let bytes = include_bytes!("resources/cursor2.rgba");
+    let udev
+        = UdevBackend::new(&mut event_loop, &context, session.clone(), UdevHandlerImpl {
+            shell_state_token,
+            compositor_token,
+            window_map: window_map.clone(),
+            pointer_location: pointer_location.clone(),
+            pointer_image: ImageBuffer::from_raw(64, 64, bytes.to_vec()).unwrap(),
+            logger: log.clone(),
+        }, log.clone()).unwrap();
+
+    let udev_token = event_loop.state().insert(udev);
+    let udev_session_id = notifier.register(udev_token.clone());
+
+    let (seat_token, _) = Seat::new(&mut event_loop, session.seat().into(), log.clone());
+
+    let pointer = event_loop.state().get_mut(&seat_token).add_pointer();
+    let keyboard = event_loop
+        .state()
+        .get_mut(&seat_token)
+        .add_keyboard("", "", "", None, 1000, 500)
+        .expect("Failed to initialize the keyboard");
+
+    let (output_token, _output_global) = Output::new(
+        &mut event_loop,
+        "Drm".into(),
+        PhysicalProperties {
+            width: 0,
+            height: 0,
+            subpixel: wl_output::Subpixel::Unknown,
+            maker: "Smithay".into(),
+            model: "Generic DRM".into(),
+        },
+        log.clone(),
+    );
+
+    let (w, h) = (1920, 1080); // Hardcode full-hd res
+    event_loop
+        .state()
+        .get_mut(&output_token)
+        .change_current_state(
+            Some(Mode {
+                width: w as i32,
+                height: h as i32,
+                refresh: 60_000,
+            }),
+            None,
+            None,
+        );
+    event_loop
+        .state()
+        .get_mut(&output_token)
+        .set_preferred(Mode {
+            width: w as i32,
+            height: h as i32,
+            refresh: 60_000,
+        });
+
+    /*
+     * Initialize libinput backend
+     */
+    let seat = session.seat();
+    let mut libinput_context = Libinput::new_from_udev::<LibinputSessionInterface<Arc<Mutex<DirectSession>>>>(session.into(), &context);
+    let libinput_session_id = notifier.register(libinput_context.clone());
+    libinput_context.udev_assign_seat(&seat);
+    let mut libinput_backend = LibinputInputBackend::new(libinput_context, log.clone());
+    libinput_backend.set_handler(LibinputInputHandler {
+        log: log.clone(),
+        pointer,
+        keyboard,
+        window_map: window_map.clone(),
+        pointer_location,
+        screen_size: (w, h),
+        serial: 0,
+        running: running.clone(),
+    });
+    let libinput_event_source = libinput_bind(libinput_backend, &mut event_loop).unwrap();
+
+    let session_event_source = direct_session_bind(notifier, &mut event_loop, log.clone()).unwrap();
+    let udev_event_source = udev_backend_bind(&mut event_loop, udev_token).unwrap();
```

Comment:
Is it necessary to have this "bind" step, separated from the creation of these backends? Couldn't the backend just take the event loop as argument at creating and do this internally?

Reply:
Coming back to this after reading more of the PR, I can elaborate a little. In my opinion, processes like "get this value, then store it in the state event loop, then give me back the token" are not very ergonomic:

- either the value must absolutely be in the event loop state (as in there is no reasonable way to use them in any other way), in which case the init functions should insert it themselves
- or the value could possibly be used in an other way, in which case we should not force the state token use.

But overall, if the user cannot do anything with this value other than store it in the state and give me back the token, they should not see the value at all.

This is what I've been leaning to with the implementation/implementation data schemes in the wayland-side implementation: if my implementation data needs to be accessed by other things outside of my module (and specifically I don't know in advance what or how), I shove it in the state and use a state token. But if it is only used internally by the module, I try to keep it as implementation data, hidden from the user (note that using an internal state token to ease sharing works too, but it's not clear how resource freeing would be handled in this case).

Though, I'm just opening a discussion here, not saying you should change everything. I lack most of the context for how you designed this whole API.

Reply:
I think this whole comment is still valid overall.